lib/espree.js

```diff
@@ -268,11 +272,15 @@ module.exports = () => Parser => {
                 if (result.range) {
                     result.range[0]--;
                     result.range[1] += (terminalDollarBraceL ? 2 : 1);
+                    result.start = result.range[0];
+                    result.end = result.range[1];
                 }
 
                 if (result.loc) {
                     result.loc.start.column--;
                     result.loc.end.column += (terminalDollarBraceL ? 2 : 1);
+                    result.start = result.loc.start.column;
+                    result.end = result.loc.end.column;
```

Comment:
Same as the comments above.

Also, I'm not really sure is it safe to modify Acorn's `start` and `end` at this point.

Reply:
how about 

```js
                result.start--;
                result.end += (terminalDollarBraceL ? 2 : 1);
```
 and removing the current diff

Reply:
That looks good!

My only concern is - does Acorn maybe use `start` and/or `end` from finished nodes for some parsing logic that happens afterwards, so could altering these values _at this point_ maybe break something?

Reply:
> My only concern is - does Acorn maybe use start and/or end from finished nodes for some parsing logic that happens afterwards, so could altering these values at this point maybe break something?


not sure about but I think it shouldnt break but cant confirm that as of now

Reply:
There are some lines in acorn where it seems that a previously set `start`/`end` is being used, like [this line](https://github.com/acornjs/acorn/blob/85ce0a50eb3fd1938e0ad49391b6936043136d41/acorn/src/expression.js#L180) and [this line](https://github.com/acornjs/acorn/blob/85ce0a50eb3fd1938e0ad49391b6936043136d41/acorn/src/expression.js#L279). I'm not familiar with acorn code to be sure if that's really the case, and whether modifying `start`/`end` on this particular node type (`TemplateElement`) in `finishNode()` could really affect acorn's parsing, now or in the future.

We could maybe collect `TemplateElement` nodes in an array, and postpone modifying `start` and `end` to the place where we modify `Program` (after `super.parse()`, I believe the ast is completely finished at that point). This would add an extra array, a loop, and a bit of complexity, so I'm not sure if it's necessary or maybe overly cautious.

@nzakas what do you think about this?

Reply:
I *think* it’s okay to modify TemplateElement here. The places where Acorn uses this info doesn’t appear to be related to template strings at all.

If we want to be cautious, I think your suggestion of collecting the nodes and applying the changes later makes sense. 

I guess the question we need to answer is how confident are we that Acorn will never need the original position info?

Reply:
> We could maybe collect `TemplateElement` nodes in an array, and postpone modifying `start` and `end` to the place where we modify `Program` (after `super.parse()`, I believe the ast is completely finished at that point). This would add an extra array, a loop, and a bit of complexity, so I'm not sure if it's necessary or may be overly cautious.

If I am not wrong here, then we would need to copy the whole nodes to the templateElement in order to update the ast before returning the program, and also we might need to traverse from `program` as well right in order to update!

> I guess the question we need to answer is how confident are we that Acorn will never need the original position info?

tough one. not sure about this.

Reply:
> I guess the question we need to answer is how confident are we that Acorn will never need the original position info?

Since it seems that Acorn does sometimes use `start`/`end` from finished nodes, we probably can't be sure that some future version won't need `start`/`end` values from these particular nodes and use them in a way that might break if we modify them, so I'm leaning towards being cautious unless it would significantly affect performance.

Reply:
> If I am not wrong here, then we would need to copy the whole nodes to the templateElement in order to update the ast before returning the program, and also we might need to traverse from `program` as well right in order to update!

In ` [ESPRIMA_FINISH_NODE](result)`, if  `result.type === "TemplateElement"` we could push the `result` reference to an array, and later iterate over that array in `parse()` to update the nodes.

Reply:
> If I am not wrong here, then we would need to copy the whole nodes to the templateElement in order to update the ast before returning the program, and also we might need to traverse from `program` as well right in order to update!

In ` [ESPRIMA_FINISH_NODE](result)`, if  `result.type === "TemplateElement"` we could push the `result` reference to an array, and later iterate over that array in `parse()` to update the nodes.

Reply:
That sounds like a good plan to me.

Reply:
> > If I am not wrong here, then we would need to copy the whole nodes to the templateElement in order to update the ast before returning the program, and also we might need to traverse from `program` as well right in order to update!
> 
> In ` [ESPRIMA_FINISH_NODE](result)`, if `result.type === "TemplateElement"` we could push the `result` reference to an array, and later iterate over that array in `parse()` to update the nodes.

just to confirm if I am getting this or not, we would still need two loops - one for traversing the AST (`program`) and one for the array for the `templateElements` in `parse` method right?

Reply:
Only one loop in `parse()`, I believe there is no need to traverse AST.

` [ESPRIMA_FINISH_NODE](result)` is called for each AST node, so we can do something like this there:

```js
if (result.type === "TemplateElement") {
    this[STATE].templateElements.push(result);
}
``` 

Then, `parse()` will have all TemplateElement nodes in `this[STATE].templateElements`.

Reply:
Yeah but I am asking about updating the `program`  in `parse` method . before returning the `program` we need to update all `templateElement` nodes right. so we need to traverse the `program` and for every `templateElement` node, we need to search the node that is stored in `this[STATE].templateElement` (need to loop this array)

Reply:
We should modify `start` and `end` on TemplateElement nodes after `super.parse()`.

All `TemplateElement` objects that are in the program's tree will be referenced from `this[STATE].templateElement`, so there is no need to search for them in the tree.

Reply:
If that causes the confusion: the top-level `espree.parse()` [always creates a new instance](https://github.com/eslint/espree/blob/3b4ca9e3141514ffac93bb7fef6c1329370df310/espree.js#L134) of [Espree](https://github.com/eslint/espree/blob/3b4ca9e3141514ffac93bb7fef6c1329370df310/lib/espree.js#L51). Therefore, `this[STATE].templateElements` will contain only `TemplateElement` nodes from that `program`'s tree.

In [parse()](https://github.com/eslint/espree/blob/3b4ca9e3141514ffac93bb7fef6c1329370df310/lib/espree.js#L142), we just need to modify `start` and `end` in each element of `this[STATE].templateElements` array. That should be done somewhere after `super.parse()`.

At that point, all TemplateElement objects from the tree are already in `this[STATE].templateElements`, there's nothing else in `this[STATE].templateElements`, and those are the same objects as the objects in the tree, so there is no need to traverse the tree.